src/OperationResponse.php

```diff
@@ -241,9 +299,9 @@ public function getResult()
      * @return Status|null The status of the operation in case of failure, or null if
      *                                 operationFailed() is false.
      */
-    public function getError()
+    public function getError(Message $lastProtoResponse = null)
```

Comment:
Nit: I don't see use of the `operation_field = ERROR_MESSAGE` or `operation_field = ERROR_CODE` configurations. Do we need to use these? FWIW I think the `Operation` in `compute` has an `Error error` field so this might not be necessary.

Reply:
Currently ee just return the raw error object to our customers to inspect, and don't currently provide any convenience functions around it.

We could do something like add a `getErrorMessage` and `getErrorCode` method on the `OperationResponse` object. That way users could handle operations from Compute and OnePlatform the same. But this could be added at any time.

@dwsupplee what do you think?

Reply:
> Currently we just return the raw error object to our customers to inspect, and don't currently provide any convenience functions around it.

Right. But does it make sense for the `getError` method to construct an error using the `Operation` fields `error_code` and `error_message` if the `Operation` proto doesn't have an `error` field? Feel like these annotations must have _some_ use since they are there. As you said, we could add something later for that.

Reply:
The other operation responses returns a `Google\Rpc\Status` object (see [here](https://github.com/googleapis/common-protos-php/blob/master/src/Rpc/Status.php)) on calls to `$lastProtoResponse->getError()`. We could construct one of these, which feels _slightly_ hacky to me, since that's not actually what the API returns. But it could be done.

Is that what the other languages are doing?

Reply:
> which feels _slightly_ hacky to me, since that's not actually what the API returns

I agree, that's kind of hacky...IDK what other languages are doing, but maybe we should consult with Vadym about the purpose of the annotation.

Reply:
@vam-google

Reply:
**tl;dr;** I guess there are two parts here:
1) Do we need those annotations to do polling right?
2) Do we need those annotations to expose error details to the users in a well-defined form (i.e. do some wrapping around raw objects to  make standard and non-standard error details look the same on the surface).

The first question is critical because it is about correctness (i.e. required). The second one is less important, because it is about usability (optional, and you may choose to not do it, if you think is better not do it).

More details below:

Whatever PHP does here the following functionality should be present at minimum in the generated gapic library:
1) Users should be able to get error details in case if error occurs during polling.

2) If polling is done automatically, the polling logic should be able to take reasonable actions in case of an error (usually it just means stop polling and report an error in a langauge-idiomatic way). To take the reasonable action, the logic needs to know where the error information is supposed to come from - this is why the annotations exist in the first place.

3) If polling is not done automatically, then please clarify what exactly LRO support does and what it doesn't do for the user in PHP. To the best of my understanding polling is started by user explicitly in PHP, but then it is done automatically, and to properly stop in case of an error, we need that info from the operation. Note, the polling request may potentially return 200 OK as HTTP error code, but the returned object itself may contain error details if polling didn't go well (i.e. http error codes and LRO polling are on different layers - success on rpc layer does not mean success on LRO logic layer, which is built on top of that rpc). 

4) If in PHP the standard Operation object is not wrapped in some return object-independent abstraction, and the same approach was chosen for DIREGAPIC lro (i.e. the custom Operation object is still exposed raw to the user), then, I guess there is no other option for the error part of the object - it must be exposed raw too. I.e. if the whole object exposed as is the error details fields of it have to be exposed as is as well. So in that case, yes, you do not need extra annotations to reason about error information in the object, because the object as a whole is not processed/wrapped in a special way. I would guess that the error details are still needed to properly stop pppolling in case of an error. If that is not the case, please clarify why it is not true for PHP.

As for constructing fake `Google\Rpc\Status` object I would advice against it. Basically the idea is to minimize numbers of "fakes"  and not try to make API look different from what it actually is, if possible. Sometimes it may be unavoidable, if "fake" intermediate objects are needed to make the rest of LRO machinery work, but for cases when it can be avoided, please do not construct unnecessary facade objects.


On the other hand, form Brent's response:

> We could do something like add a getErrorMessage and getErrorCode method on the OperationResponse object. That way users could handle operations from Compute and OnePlatform the same. But this could be added at any time.

I think it has value in it, but as long as the unified interface is generic and does not make one API pretend that it is the other one (i.e. make compute LRO pretend like they return Rpc/Status  while it is not). If both standard and non/standard error responses are wrapped in a generic abstraction (and neither one tries to pretend that it is the other one), this is what is exposed to the user, and that can be done in a backward compatible way - i believe that would be the best option.